### PR TITLE
[SQL Server] (No.3) add sql server creation ui codes - 2

### PR DESCRIPTION
**Pull request by shenqianjin**
image

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+package com.microsoft.azure.toolkit.intellij.sqlserver.creation;
+
+import com.microsoft.azure.management.resources.Subscription;
+import com.microsoft.azure.toolkit.intellij.appservice.resourcegroup.ResourceGroupComboBox;
+import com.microsoft.azure.toolkit.intellij.appservice.subscription.SubscriptionComboBox;
+import com.microsoft.azure.toolkit.intellij.common.AzureFormPanel;
+import com.microsoft.azure.toolkit.intellij.common.AzurePasswordFieldInput;
+import com.microsoft.azure.toolkit.intellij.common.TextDocumentListenerAdapter;
+import com.microsoft.azure.toolkit.intellij.mysql.AdminUsernameTextField;
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:14]: consider rename the packages of these common classes

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            public void onDocumentChanged() {
+                if (!adminUsernameTextField.isValueInitialized()) {
+                    adminUsernameTextField.setValueInitialized(true);
+                }
+            }
+        };
+    }
+
+    private void onSecurityAllowAccessFromAzureServicesCheckBoxChanged(final ItemEvent e) {
+        config.setAllowAccessFromAzureServices(e.getStateChange() == ItemEvent.SELECTED);
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:96]: As we have called `config.setAllowAccessFromAzureServices(security.getAllowAccessFromAzureServicesCheckBox().isSelected())` in `getData()`, do we need to add extra listener to change the config value?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java (hunk 1 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:14]: Let us keep it in mind and try to do it when we are not busy

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java (hunk 2 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:96]: removing these lines is good and should be ok.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    @Override
+    public SqlServerConfig getData() {
+        config.setServerName(serverNameTextField.getText());
+        config.setAdminUsername(adminUsernameTextField.getText());
+        config.setPassword(passwordField.getPassword());
+        config.setConfirmPassword(confirmPasswordField.getPassword());
+        config.setSubscription(subscriptionComboBox.getValue());
+        config.setResourceGroup(resourceGroupComboBox.getValue());
+        config.setRegion(regionComboBox.getValue());
+        /*if (StringUtils.isNotBlank(versionComboBox.getValue())) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:108]: delete this code

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        if (config.getSubscription() != null) {
+            subscriptionComboBox.setValue(config.getSubscription());
+        }
+        if (config.getResourceGroup() != null) {
+            resourceGroupComboBox.setValue(config.getResourceGroup());
+        }
+        if (config.getRegion() != null) {
+            regionComboBox.setValue(config.getRegion());
+        }
+        /*if (config.getVersion() != null) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:139]: delete this?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    private void initListeners() {
+        this.subscriptionComboBox.addItemListener(this::onSubscriptionChanged);
+        this.adminUsernameTextField.getDocument().addDocumentListener(generateAdminUsernameListener());
+    }
+
+    private void onSubscriptionChanged(final ItemEvent e) {
+        if (e.getStateChange() == ItemEvent.SELECTED && e.getItem() instanceof Subscription) {
+            final Subscription subscription = (Subscription) e.getItem();
+            this.resourceGroupComboBox.setSubscription(subscription);
+            this.serverNameTextField.setSubscription(subscription);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:77]: setSubscriptionId

Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:77]: updated.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    private void onSubscriptionChanged(final ItemEvent e) {
+        if (e.getStateChange() == ItemEvent.SELECTED && e.getItem() instanceof Subscription) {
+            final Subscription subscription = (Subscription) e.getItem();
+            this.resourceGroupComboBox.setSubscription(subscription);
+            this.serverNameTextField.setSubscriptionId(subscription.subscriptionId());
+            this.regionComboBox.setSubscription(subscription);
+        }
+    }
+
+    private DocumentListener generateAdminUsernameListener() {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:83]: no need to separate a method generateAdminUsernameListener

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        passwordFieldInput = PasswordUtils.generatePasswordFieldInput(this.passwordField, this.adminUsernameTextField);
+        confirmPasswordFieldInput = PasswordUtils.generateConfirmPasswordFieldInput(this.confirmPasswordField, this.passwordField);
+    }
+
+    private void initListeners() {
+        this.subscriptionComboBox.addItemListener(this::onSubscriptionChanged);
+        this.adminUsernameTextField.getDocument().addDocumentListener(generateAdminUsernameListener());
+    }
+
+    private void onSubscriptionChanged(final ItemEvent e) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:74]: no need to extract a method

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java (hunk 6 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:83]: it's only a private method. I think it should be ok.
If not, the init() method will have a lot of code lines, isn't it?

Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:83]: merge all listeners into only one method? I think it not a good idea.

image

Separate it to two methods is better than seting them in one method. isn't it ?
and It maybe add more in the future.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java (hunk 7 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java:74]: same above.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java (hunk 6 above)
Comment (shenqianjin) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/sqlserver/creation/SqlServerCreationAdvancedPanel.java]: renamed.

**Review by andxu — APPROVED** [review]
Please refactor the code, replacing Subscription and Resource Group with new models.

**Pull request merged**